neuralmonkey/encoders/imagenet_encoder.py

```diff
@@ -97,56 +98,70 @@ def __init__(self,
         self._network_type = network_type
-        self.input_plc = tf.placeholder(
-            tf.float32, [None, self.HEIGHT, self.WIDTH, 3])
+        self._attention_layer = attention_layer
+        self._encoded_layer = encoded_layer
+        self._fine_tune = fine_tune
 
-        if network_type not in SUPPORTED_NETWORKS:
+        if self._network_type not in SUPPORTED_NETWORKS:
             raise ValueError(
                 "Network '{}' is not among the supoort ones ({})".format(
-                    network_type, ", ".join(SUPPORTED_NETWORKS.keys())))
+                    self._network_type, ", ".join(SUPPORTED_NETWORKS.keys())))
 
-        scope, net_function = SUPPORTED_NETWORKS[network_type]
+        scope, net_function = SUPPORTED_NETWORKS[self._network_type]
         with tf_slim.arg_scope(scope()):
-            _, end_points = net_function(self.input_plc)
-
-        with tf.variable_scope(self.name):
-            if attention_layer is not None:
-
-                if attention_layer not in end_points:
-                    raise ValueError(
-                        "Network '{}' does not contain endpoint '{}'".format(
-                            network_type, attention_layer))
-
-                net_output = end_points[attention_layer]
-
-                if len(net_output.get_shape()) != 4:
-                    raise ValueError(
-                        ("Endpoint '{}' for network '{}' cannot be "
-                         "a convolutional map, its dimensionality is: {}."
-                        ).format(attention_layer, network_type,
-                                 ", ".join([str(d.value) for d in
-                                            net_output.get_shape()])))
-
-                if not fine_tune:
-                    net_output = tf.stop_gradient(net_output)
-                # pylint: disable=no-member
-                shape = [s.value for s in net_output.get_shape()[1:]]
-                # pylint: enable=no-member
-                self.__attention_tensor = tf.reshape(
-                    net_output, [-1, shape[0] * shape[1], shape[2]])
-
-            if encoded_layer is not None:
-                if encoded_layer not in end_points:
-                    raise ValueError(
-                        "Network '{}' does not contain endpoint '{}'.".format(
-                            network_type, encoded_layer))
-
-                self.encoded = tf.squeeze(end_points[encoded_layer], [1, 2])
-                if not fine_tune:
-                    self.encoded = tf.stop_gradient(self.encoded)
-            else:
-                self.encoded = tf.reduce_mean(net_output, [1, 2])
-    # pylint: enable=too-many-arguments,too-many-locals
+            _, self._end_points = net_function(self.input_image)
+
+    @tensor
+    def input_image(self) -> tf.Tensor:
+        return tf.placeholder(
+            tf.float32, [None, self.HEIGHT, self.WIDTH, 3])
+
+    @tensor
+    def cnn_states(self) -> tf.Tensor:
```

Comment:
jestli to někdy vrací None, tak by to mělo bejt nějak zohledněný v tý anotaci, ne?